reference/src/functions/macros.md

```diff
@@ -0,0 +1,680 @@
+# Macro Functions
+
+Macro functions are a way of defining functions that are expanded at compile time at each call site.
+The arguments of the macro are not evaluated eagerly like a normal function, and instead are
+substituted by expression. In addition, the caller can supply code to the macro via
+[lambdas](#lambdas).
+
+These expression substitution mechanics make `macro` functions similar
+[to macros found in other programming languages](<https://en.wikipedia.org/wiki/Macro_(computer_science)>);
+however, they are more constrained in Move than you might expect from other languages. The parametrs
+and return values of `macro` functions are still typed--though this can be partially relaxed with
+the [`_` type](TODO). The upside of this restriction however, is that `macro` functions can be used
+anywhere a normal function can be used, which is notably helpful with
+[method syntax](../method-syntax.md).
+
+A more extensive
+[syntactic macro](<https://en.wikipedia.org/wiki/Macro_(computer_science)#Syntactic_macros>) system
+may come in the future.
+
+## Syntax
+
+`macro` functions have a similar syntax to normal functions. However, all type parameter names and
+all parameter names must start with a `$`.
+
+```text
+<visibility>? macro fun <identifier><[$type_parameters: constraint],*>([$identifier: type],*): <return_type> <function_body>
+```
+
+For example, the following `macro` function takes a vector and a lambda, and applies the lambda to
+each element of the vector to construct a new vector.
```

Comment:
Consider moving this example to the top, so someone consulting the reference sees it first thing.